### PR TITLE
Add null check for channel when generating error message

### PR DESCRIPTION
**Pull request by sigurjonviktorsson**
## Motivation and Context

I think this issue is relevant https://github.com/aws/aws-sdk-java-v2/issues/3435

There's a possibility of channel being null if the channel acquisition fails in this code path:
https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyRequestExecutor.java#L175

```
2022-11-29 13:35:37,629  WARN  [aws-java-sdk-NettyEventLoop-1-0] io.netty.channel.AbstractChannelHandlerContext - An exception 'java.lang.NullPointerException' [enable DEBUG level for full stacktrace] was thrown by a user handler's exceptionCaught() method while handling the following exception:
software.amazon.awssdk.http.nio.netty.internal.FutureCancelledException: software.amazon.awssdk.core.exception.ApiCallAttemptTimeoutException: HTTP request execution did not complete before the specified timeout configuration: 80 millis
        at software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.lambda$null$3(NettyRequestExecutor.java:136)
        at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
        at io.netty.util.concurrent.PromiseTask.run(PromiseTask.java:106)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: software.amazon.awssdk.core.exception.ApiCallAttemptTimeoutException: HTTP request execution did not complete before the specified timeout configuration: 80 millis
        at software.amazon.awssdk.core.exception.ApiCallAttemptTimeoutException$BuilderImpl.build(ApiCallAttemptTimeoutException.java:88)
        at software.amazon.awssdk.core.exception.ApiCallAttemptTimeoutException.create(ApiCallAttemptTimeoutException.java:38)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$setupAttemptTimer$5(MakeAsyncHttpRequestStage.java:214)
        at software.amazon.awssdk.core.internal.http.timers.AsyncTimeoutTask.run(AsyncTimeoutTask.java:52)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        ... 1 common frames omitted
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by sigurjonviktorsson** [comment]
@zoewangg How can we move this forward?

**Review by zoewangg — DISMISSED** [review]
LGTM. Thank you for your contribution!

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java [hunk 1]
```diff
@@ -133,9 +133,11 @@ private static String getMessageForTooManyAcquireOperationsError() {
     }
 
     public static String closedChannelMessage(Channel channel) {
-        ChannelDiagnostics channelDiagnostics = channel.attr(CHANNEL_DIAGNOSTICS).get();
-        ChannelDiagnostics parentChannelDiagnostics = channel.parent() != null ? channel.parent().attr(CHANNEL_DIAGNOSTICS).get()
-                                                                               : null;
+        ChannelDiagnostics channelDiagnostics, parentChannelDiagnostics;
```
Comment (zoewangg) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java:136]: Looks like compilation failed.
```
ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project netty-nio-client: Compilation failure: Compilation failure:

NettyUtils.java:[148,12] error: variable channelDiagnostics might not have been initialized
NettyUtils.java:[151,16] error: variable parentChannelDiagnostics might not have been initialized

```

Comment (sigurjonviktorsson) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java]: Should be fixed now.

Comment (zoewangg) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java]: I think checkstyle still failed. Are you able to build it locally?

Comment (sigurjonviktorsson) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java]: How do I run checkstyle?

`mvn install` is failing locally on something unrelated I think:
image

**Comment by sigurjonviktorsson** [comment]
@zoewangg Can you take another look?

**Comment by zoewangg** [comment]
Apologies for the lack of response. I think checkstyle still failed (`Each variable declaration must be in its own statement. [MultipleVariableDeclarations]`). The variable should be declared in separate statements as below.

```
        ChannelDiagnostics channelDiagnostics = null;
        ChannelDiagnostics parentChannelDiagnostics = null;
```

What JDK version are you using to build the SDK? I think there are some false positive checkstyle issues when building with JDK 18+ and we will work on fixing it.

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java [hunk 2]
```diff
@@ -133,9 +133,11 @@ private static String getMessageForTooManyAcquireOperationsError() {
     }
 
     public static String closedChannelMessage(Channel channel) {
-        ChannelDiagnostics channelDiagnostics = channel.attr(CHANNEL_DIAGNOSTICS).get();
```
Comment (joviegas) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java:136]:
```java

        ChannelDiagnostics channelDiagnostics = channel != null && channel.attr(CHANNEL_DIAGNOSTICS) != null ?
                                                channel.attr(CHANNEL_DIAGNOSTICS).get() : null;
        ChannelDiagnostics parentChannelDiagnostics = channel != null && channel.parent() != null ?
                                                      channel.parent().attr(CHANNEL_DIAGNOSTICS).get() : null;
```

**Comment by joviegas** [comment]
Can we add Junit test cases in NettyUtilTest.java

```java
    @Test
    public void closedChannelMessage_with_nullChannelAttribute() throws Exception {

        Channel channel = Mockito.mock(Channel.class);
        when(channel.parent()).thenReturn(null);

        assertThat(NettyUtils.closedChannelMessage(channel))
            .isEqualTo(NettyUtils.CLOSED_CHANNEL_ERROR_MESSAGE);
    }

    @Test
    public void closedChannelMessage_with_nullChannel() throws Exception {
        Channel channel = null;
        assertThat(NettyUtils.closedChannelMessage(channel))
            .isEqualTo(NettyUtils.CLOSED_CHANNEL_ERROR_MESSAGE);
    }


    @Test
    public void closedChannelMessage_with_nullParentChannel() throws Exception {

        Channel channel = Mockito.mock(Channel.class);
        Attribute attribute = Mockito.mock(Attribute.class);
        when(channel.parent()).thenReturn(null);
        when(channel.attr(Mockito.any())).thenReturn(attribute);

        assertThat(NettyUtils.closedChannelMessage(channel))
            .isEqualTo(NettyUtils.CLOSED_CHANNEL_ERROR_MESSAGE);
    }

```

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java (hunk 2 above)
Comment (sigurjonviktorsson) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java:136]: That looks good to me, I added a null check for the `CHANNEL_DIAGNOSTICS` attribute on the parent as well. Not sure if this passes style rules.

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java (hunk 1 above)
Comment (joviegas) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/utils/NettyUtils.java]: There were some compilation issues with NettyUtilsTest , I have fixed and updated the PR.
The builds are passing now

**Review by zoewangg — APPROVED** [review]

**Pull request merged**